File: leetcode/medium/49_group_anagrams/group_anagrams.py
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        anagrams = {}
        for str in strs:
            anagram_dict = tuple(sorted(self.createAnagramDict(str).items()))
            if anagram_dict not in anagrams:
                anagrams[anagram_dict] = [str]
            else:
                anagrams[anagram_dict].append(str)
        result = []
        for element in anagrams.values():
            result.append(element)
        return result

    def createAnagramDict(self, str):
        char_counts = {}
        for char in str:
            if not char in char_counts:
                char_counts[char] = 1
            else:
                char_counts[char] += 1
        return char_counts


if __name__ == "__main__":
    print(Solution().groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
    print(Solution().groupAnagrams(["ddddddddddg", "dgggggggggg"]))

File: leetcode/medium/49_group_anagrams/group_anagrams.cpp
#include <algorithm>
#include <string>
#include <unordered_map>
#include <vector>

class Solution {
public:
  std::vector<std::vector<std::string>>
  groupAnagrams(std::vector<std::string> &strs) {
    std::unordered_map<std::string, std::vector<std::string>> anagram_groups;
    std::vector<std::string> keys;
    for (auto &str : strs) {
      auto key = std::string(str.begin(), str.end());
      std::sort(key.begin(), key.end());
      const auto maybe_key = anagram_groups.find(key);
      if (maybe_key == anagram_groups.end()) {
        anagram_groups.insert({key, {str}});
        keys.push_back(key);
      } else {
        maybe_key->second.push_back(str);
      }
    }
    std::vector<std::vector<std::string>> result;
    for (auto &key : keys) {
      result.push_back(anagram_groups[key]);
    }
    return result;
  }
};
